metagraph/src/cli/assemble.cpp

```diff
@@ -120,15 +224,56 @@ int assemble(Config *config) {
 
     logger->trace("Graph loaded in {} sec", timer.elapsed());
 
-    std::unique_ptr<graph::AnnotatedDBG> anno_graph;
     if (config->infbase_annotators.size()) {
-        anno_graph = initialize_annotated_dbg(graph, *config);
+        assert(config->label_mask_file.size());
+        auto anno_graph = initialize_annotated_dbg(graph, *config);
 
-        logger->trace("Masking graph...");
+        logger->trace("Generating masked graphs...");
 
-        graph = mask_graph(*anno_graph, config);
+        std::filesystem::remove(
+            utils::remove_suffix(config->outfbase, ".gz", ".fasta") + ".fasta.gz"
+        );
+
+        std::mutex file_open_mutex;
+        std::mutex write_mutex;
+
+        size_t num_parallel_assemblies = std::max(
+            1u, std::min(config->parallel_assemblies, get_num_threads())
+        );
+        size_t num_threads_per_traversal = std::max(
+            size_t(1), get_num_threads() / num_parallel_assemblies
+        );
 
-        logger->trace("Masked in {} sec", timer.elapsed());
+        call_masked_graphs(*anno_graph, config,
+            [&](const graph::MaskedDeBruijnGraph &graph, const std::string &header) {
+                std::lock_guard<std::mutex> file_lock(file_open_mutex);
+                seq_io::FastaWriter writer(config->outfbase, header,
+                                           config->enumerate_out_sequences,
+                                           get_num_threads() > 1, /* async write */
+                                           "a" /* append mode */);
+
+                if (config->unitigs || config->min_tip_size > 1) {
+                    graph.call_unitigs([&](const auto &unitig, auto&&) {
+                                           std::lock_guard<std::mutex> lock(write_mutex);
+                                           writer.write(unitig);
+                                       },
+                                       num_threads_per_traversal,
+                                       config->min_tip_size,
+                                       config->kmers_in_single_form);
+                } else {
+                    graph.call_sequences([&](const auto &seq, auto&&) {
+                                             std::lock_guard<std::mutex> lock(write_mutex);
+                                             writer.write(seq);
+                                         },
+                                         num_threads_per_traversal,
+                                         config->kmers_in_single_form);
+                }
+            },
```

Comment:
The is written as a critical section. So you run only one call_sequences at a time? If so, why do you use only `num_threads_per_traversal` threads and not all `get_num_threads()`?

What is the parallelization scheme here?

Reply:
The goal in the end is to have several assembly experiments running in parallel and have them all write their results to the same output file.

One approach is to just have one output file per experiment.

The other approach (keeping the single-file output) would involve either modifying FastaWriter to have a header associated with each sequence, or do the current approach where each assembly that finishes locks the output files to append.

Reply:
Are you sure we need these two levels of parallelization?

Why do we need to run multiple assemblies at the same time, if each assembly already happens in parallel?

Reply:
Since different assembly experiments are independent, it scales better than doing one with multiple threads

Reply:
Though this is theoretical, I don't remember if I actually measured it to see if it makes a difference

Reply:
Thinking about this again, it's probably not worth it. If two assemblies are done at the same time, then you have to lock and you end up using far fewer threads than you intended